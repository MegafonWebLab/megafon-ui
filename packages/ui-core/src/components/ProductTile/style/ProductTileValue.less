@import '~styles/base.less';

@b: e('.mfui-product-tile-value');

@{b} {
    position: relative;
    white-space: nowrap;
    display: inline-block;
    text-align: center;
    overflow: hidden;
    vertical-align: bottom;
    max-height: 28px;

    @media @mobileB {
        max-height: 32px;
    }

    @media @desktopS {
        max-height: 36px;
    }

    @media @desktopMU {
        max-height: 40px;
    }

    &__prev {
        opacity: 1;
        transition: none;
    }
    &__prev_hide {
        opacity: 0;
        transition: opacity 0.5s;
        position: absolute;
        left: 50%;
        transform: translateX(-50%);
    }

    &__current {
        opacity: 1;
        transition: opacity 0.5s;
        display: inline-block;
    }
    &__current_hide {
        opacity: 0;
        transition: none;
        position: absolute;
    }
}
