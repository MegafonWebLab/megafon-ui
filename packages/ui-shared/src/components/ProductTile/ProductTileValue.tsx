import * as React from 'react';
import * as PropTypes from 'prop-types';
import { cnCreate } from '@megafon/ui-core';
import './style/ProductTileValue.less';

interface IProductTileValueProps {
    /** Value */
    value: string;
    /** hAlign */
    hAlign?: 'center';
}

interface IProductTileValueState {
    prevValue: string;
    currentValue: string;
    isAnimating: boolean;
}

const cn = cnCreate('mfui-product-tile-value');
class ProductTileValue extends React.PureComponent<IProductTileValueProps, IProductTileValueState> {
    static propTypes = {
        value: PropTypes.string.isRequired,
        hAlign: PropTypes.oneOf(['center']),
    };

    constructor(props: IProductTileValueProps) {
        super(props);

        this.state = {
            prevValue: '',
            currentValue: props.value,
            isAnimating: false,
        };
    }

    static getDerivedStateFromProps(props: IProductTileValueProps, state: IProductTileValueState) {
        const isChangedValue = props.value !== state.currentValue;

        return {
            prevValue: isChangedValue ? state.currentValue : state.prevValue,
            currentValue: props.value,
            isAnimating: isChangedValue || state.isAnimating,
        };
    }

    componentDidUpdate() {
        const ANIMATION_DELAY = 50;

        if (typeof window === 'undefined') {
            return;
        }

        window.setTimeout(() => {
            this.setState({ isAnimating: false });
        }, ANIMATION_DELAY);
    }

    renderPrevPrice() {
        const { prevValue, isAnimating } = this.state;

        return (
            <span className={cn('prev', { hide: !isAnimating })}>
                {prevValue}
            </span>
        );
    }

    renderCurrentPrice() {
        const { currentValue, isAnimating } = this.state;

        return (
            <span className={cn('current', { hide: isAnimating })}>
                {currentValue}
            </span>
        );
    }

    render() {
        const { hAlign } = this.props;

        return (
            <span className={cn('', { 'h-align': hAlign })}>
                {this.renderPrevPrice()}
                {this.renderCurrentPrice()}
            </span>
        );
    }
}

export default ProductTileValue;
