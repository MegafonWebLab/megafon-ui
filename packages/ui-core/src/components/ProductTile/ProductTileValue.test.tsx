import * as React from 'react';
import { shallow } from 'enzyme';
import ProductTileValue from './ProductTileValue';

const props = {
    value: 'werwer',
};

describe('<ProductTileValue />', () => {
    it('it renders ProductTileValue', () => {
        const wrapper = shallow(<ProductTileValue {...props} hAlign="center" />);
        expect(wrapper).toMatchSnapshot();
    });
});
