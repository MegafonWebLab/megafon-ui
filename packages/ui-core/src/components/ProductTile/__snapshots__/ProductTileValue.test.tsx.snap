// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductTileValue /> it renders ProductTileValue 1`] = `
<span
  className="mfui-product-tile-value mfui-product-tile-value_h-align_center"
>
  <span
    className="mfui-product-tile-value__prev mfui-product-tile-value__prev_hide"
  />
  <span
    className="mfui-product-tile-value__current"
  >
    werwer
  </span>
</span>
`;
